expression/builtin_json.go

```diff
@@ -51,471 +65,448 @@ var (
+	// Unquote JSON value.
+	_ builtinFunc = &builtinJSONUnquoteSig{}
+	// Create JSON array.
+	_ builtinFunc = &builtinJSONArraySig{}
+	// Create JSON object.
+	_ builtinFunc = &builtinJSONObjectSig{}
+	// Return data from JSON document.
+	_ builtinFunc = &builtinJSONExtractSig{}
+	// Insert data into JSON document.
+	_ builtinFunc = &builtinJSONSetSig{}
+	// Insert data into JSON document.
+	_ builtinFunc = &builtinJSONInsertSig{}
+	// Replace values in JSON document.
+	_ builtinFunc = &builtinJSONReplaceSig{}
+	// Remove data from JSON document.
+	_ builtinFunc = &builtinJSONRemoveSig{}
+	// Merge JSON documents, preserving duplicate keys.
+	_ builtinFunc = &builtinJSONMergeSig{}
+)
 
 type jsonTypeFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONTypeSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
 func (c *jsonTypeFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONTypeSig{newBaseBuiltinFunc(args, ctx)}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpJSON)
+	bf.tp.Flen = jsonTypeFlen
+	args[0].GetType().Decimal = decimal4CastJSONDirectly
+	sig := &builtinJSONTypeSig{baseStringBuiltinFunc{bf}}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonTypeSig)
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONTypeSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinJSONTypeSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	var j json.JSON
+	j, isNull, err = b.args[0].EvalJSON(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
 	}
-	return JSONType(args, b.ctx.GetSessionVars().StmtCtx)
+	return j.Type(), false, nil
 }
 
 type jsonExtractFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONExtractSig struct {
-	baseBuiltinFunc
+	baseJSONBuiltinFunc
 }
 
 func (c *jsonExtractFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONExtractSig{newBaseBuiltinFunc(args, ctx)}
+	argTps := make([]evalTp, 0, len(args))
+	argTps = append(argTps, tpJSON)
+	for range args[1:] {
+		argTps = append(argTps, tpString)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpJSON, argTps...)
+	args[0].GetType().Decimal = decimal4CastJSONDirectly
+	sig := &builtinJSONExtractSig{baseJSONBuiltinFunc{bf}}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonExtractSig)
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONExtractSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinJSONExtractSig) evalJSON(row []types.Datum) (res json.JSON, isNull bool, err error) {
```

Comment:
you can remove the return variable name and avoid variable declarations, like `var s string` or `var found bool`, in this function

Reply:
fixed.